packages/vite-plugin-cloudflare/src/build.ts

```diff
@@ -0,0 +1,176 @@
+import assert from "node:assert";
+import * as fs from "node:fs";
+import * as path from "node:path";
+import colors from "picocolors";
+import type { ResolvedPluginConfig } from "./plugin-config";
+import type * as vite from "vite";
+import type { Unstable_Config } from "wrangler";
+
+export function createBuildApp(
+	resolvedPluginConfig: ResolvedPluginConfig
+): (builder: vite.ViteBuilder) => Promise<void> {
+	return async (builder) => {
+		const clientEnvironment = builder.environments.client;
+		assert(clientEnvironment, `No "client" environment`);
+		const defaultHtmlPath = path.resolve(builder.config.root, "index.html");
+		const hasClientEntry =
+			clientEnvironment.config.build.rollupOptions.input ||
+			fs.existsSync(defaultHtmlPath);
+
+		if (resolvedPluginConfig.type === "assets-only") {
+			if (hasClientEntry) {
+				await builder.build(clientEnvironment);
+			} else if (getHasPublicAssets(builder.config)) {
+				await fallbackBuild(builder, clientEnvironment);
+			}
+
+			// Return early as there are no Workers to build
+			return;
+		}
+
+		const workerEnvironments = Object.keys(resolvedPluginConfig.workers).map(
+			(environmentName) => {
+				const environment = builder.environments[environmentName];
+				assert(environment, `"${environmentName}" environment not found`);
+
+				return environment;
+			}
+		);
+
+		await Promise.all(
+			workerEnvironments.map((environment) => builder.build(environment))
+		);
+
+		const { entryWorkerEnvironmentName } = resolvedPluginConfig;
+		const entryWorkerEnvironment =
+			builder.environments[entryWorkerEnvironmentName];
+		assert(
+			entryWorkerEnvironment,
+			`No "${entryWorkerEnvironmentName}" environment`
+		);
+		const entryWorkerBuildDirectory = path.resolve(
+			builder.config.root,
+			entryWorkerEnvironment.config.build.outDir
+		);
+		const entryWorkerManifest = loadViteManifest(entryWorkerBuildDirectory);
+		const importedAssetPaths = getImportedAssetPaths(entryWorkerManifest);
+
+		if (hasClientEntry) {
+			await builder.build(clientEnvironment);
+		} else if (importedAssetPaths.size || getHasPublicAssets(builder.config)) {
+			await fallbackBuild(builder, clientEnvironment);
+		} else {
+			const entryWorkerConfigPath = path.join(
+				entryWorkerBuildDirectory,
+				"wrangler.json"
+			);
+			const workerConfig = JSON.parse(
+				fs.readFileSync(entryWorkerConfigPath, "utf-8")
+			) as Unstable_Config;
+			// Remove `assets` field as there are no assets
+			workerConfig.assets = undefined;
```

Comment:
If I understand correctly this is only a problem because the wrangler.jsonc in the `assets` playground has the following:

```
"assets": { "binding": "ASSETS" }
```

and we need to remove that because there is no point in having a binding if there are no assets to binding to.

Is that correct?

In the real world, if there are no assets then the user would not add this field in the first place, right? If so, then should this actually be a user error?

Or am I missing something here?

Reply:
That's not quite it. I've changed the implementation in this PR so that the `assets` field is always included in the output `wrangler.json` at build time, regardless of whether there are assets or not. It is then removed here if there are no assets. This makes things a lot simpler and removes the need for the `client` environment to be built first. In Vite 7 we will also be able to add this check in the `buildApp` hook and determine whether another plugin has built the `client` environment.